### PR TITLE
Implement option for saliency-based blending 

### PR DESCRIPTION
**Pull request by NathalieRombeek**
Implement option of blending based on the salience weight method, which considers the intensity of the storms, so that cells with high intensity will have a stronger weight during the blending (and hence a longer lifetime). The paper describing this method is this one:

[https://journals.ametsoc.org/view/journals/wefo/30/5/waf-d-15-0057_1.xml](https://eur03.safelinks.protection.outlook.com/?url=https%3A%2F%2Fjournals.ametsoc.org%2Fview%2Fjournals%2Fwefo%2F30%2F5%2Fwaf-d-15-0057_1.xml&data=04%7C01%7Cnathalie.rombeek%40wur.nl%7C979e24128f3c41390fa308d9d5e883bf%7C27d137e5761f4dc1af88d26430abb18f%7C0%7C0%7C637776019258539574%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000&sdata=AAOpk2uc2duwEtGIlcClEQHjnZ5hjS517W%2F6TKWY6Xw%3D&reserved=0)

### CODE REVIEW

File: pysteps/blending/interface.py [hunk 1]
```diff
@@ -32,6 +35,9 @@ def get_method(name):
     | linear_blending  | the linear blending of a nowcast method with other   |
     |                  | data (e.g. NWP data).                                |
     +------------------+------------------------------------------------------+
+    | salient_blending | the salienct blending of a nowcast method with other |
```
Comment (dnerini) [inline pysteps/blending/interface.py]: there is a typo in the description ("salienct")

also, can you please reference the original paper? (as in the description of steps, below) and perhaps add a summary sentence describing the main purpose of saliency ? just to give a rough idea of what this method does without having to read the paper

File: pysteps/blending/linear_blending.py [hunk 2]
```diff
@@ -83,6 +85,8 @@ def forecast(
       no data mask of the nowcast.
     nowcast_kwargs: dict, optional
       Dictionary containing keyword arguments for the nowcast method.
+    saliency: bool, optional
+      Standard value is False. If True, saliency will be used for blending
```
Comment (dnerini) [inline pysteps/blending/linear_blending.py]: Also here would be nice to add some context and references to the original paper.

File: pysteps/blending/linear_blending.py [hunk 3]
```diff
@@ -180,24 +184,70 @@ def forecast(
             # Set weights at times before start_blending and after end_blending
             if weight_nwp < 0.0:
                 weight_nwp = 0.0
+                if n_ens_members_max == 1:
+                    R_blended[i, :, :] = precip_nowcast[i, :, :]
```
Comment (dnerini) [inline pysteps/blending/linear_blending.py]: can we rename `R_blended` to `precip_blended`?

File: pysteps/blending/linear_blending.py [hunk 4]
```diff
@@ -180,24 +184,70 @@ def forecast(
...
             else:
-                R_blended[:, i, :, :] = (
-                    weight_nwp * precip_nwp[:, i, :, :]
-                    + weight_nowcast * precip_nowcast[:, i, :, :]
-                )
+                # Calculate weight_nowcast
+                weight_nowcast = 1.0 - weight_nwp
+
+                # Calculate output by combining precip_nwp and precip_nowcast,
+                # while distinguishing between ensemble and non-ensemble methods
```
Comment (dnerini) [inline pysteps/blending/linear_blending.py:212]: I feel this should not be necessary, but I don't have an obvious answer yet, will think about it!

File: pysteps/blending/linear_blending.py [hunk 5]
```diff
@@ -180,24 +184,70 @@ def forecast(
...
+                        R_blended[i, :, :] = (
+                            get_ws(
+                                weight_nowcast,
+                                precip_nowcast[i, :, :],
+                                precip_nwp[i, :, :],
+                            )
+                            * precip_nowcast[i, :, :]
+                            + (
+                                1
+                                - get_ws(
```
Comment (dnerini) [inline pysteps/blending/linear_blending.py]: you're computing twice the same weights, what about adding a call to `get_ws` before applying the blending? for example

```
ws = get_ws(...)
R_blended = ws * precip_nowcast + (1 - ws) * precip_nwp
```

File: pysteps/blending/linear_blending.py [hunk 6]
```diff
@@ -208,3 +258,70 @@ def forecast(
         R_blended = precip_nowcast
 
     return R_blended
+
+
+def get_ranked_salience(precip_nowcast, precip_nwp):
```
Comment (dnerini) [inline pysteps/blending/linear_blending.py]: should this be a private method? or do we want to expose it to the users?

File: pysteps/blending/linear_blending.py [hunk 7]
```diff
@@ -208,3 +258,70 @@ def forecast(
...
+    """
+    Calculate ranked salience
+
+    Args:
```
Comment (dnerini) [inline pysteps/blending/linear_blending.py]: please use the numpy docstring format here (as in the `forecast` method above)

https://numpydoc.readthedocs.io/en/latest/format.html

File: pysteps/blending/linear_blending.py [hunk 8]
```diff
@@ -208,3 +258,70 @@ def forecast(
         R_blended = precip_nowcast
 
     return R_blended
+
+
+def get_ranked_salience(precip_nowcast, precip_nwp):
+    """
+    Calculate ranked salience
```
Comment (dnerini) [inline pysteps/blending/linear_blending.py]: a slightly more verbose description would be nice ;)

File: pysteps/blending/linear_blending.py [hunk 9]
```diff
@@ -208,3 +258,70 @@ def forecast(
...
+    precip_nowcast: array_like
+      Array of shape (m,n) containing the extrapolated precipitation field at a specified timestep
+    precip_nwp: array_like
+      Array of shape (m,n) containing the NWP fields at a specified timestep
+
+    Returns:
+      Array of shape (m,n) containing ranked salience
+    """
+
+    # calcutate normalized intensity
```
Comment (dnerini) [inline pysteps/blending/linear_blending.py:263]: here we assume that the input fields are bounded at zero, correct (no negative values)?

File: pysteps/blending/linear_blending.py [hunk 10]
```diff
@@ -208,3 +258,70 @@ def forecast(
...
+
+    diff = norm_nowcast - norm_nwp
+
+    # get cumulative distribution
+    ranked_salience = norm.cdf(diff)
+
+    return ranked_salience
+
+
+def get_ws(weight, precip_nowcast, precip_nwp):
```
Comment (dnerini) [inline pysteps/blending/linear_blending.py]: again, private method perhaps?

File: pysteps/blending/linear_blending.py [hunk 11]
```diff
@@ -208,3 +258,70 @@ def forecast(
...
+    """
+    Calculate salience weight based on linear weight and ranked salience
+
+    Args:
```
Comment (dnerini) [inline pysteps/blending/linear_blending.py]: again, use numpy docstring format

File: pysteps/blending/linear_blending.py [hunk 12]
```diff
@@ -208,3 +258,70 @@ def forecast(
...
+
+    # get cumulative distribution
+    ranked_salience = norm.cdf(diff)
+
+    return ranked_salience
+
+
+def get_ws(weight, precip_nowcast, precip_nwp):
+    """
+    Calculate salience weight based on linear weight and ranked salience
```
Comment (dnerini) [inline pysteps/blending/linear_blending.py]: add reference to paper?

**Review by dnerini — COMMENTED** [review]
hello Nathalie, thanks a lot for this excellent contribution!
I really like that this comes as an option to an existing method.

In general I think this PR is already quite ready, I only left few, mostly minor comments. In particular:
Would be helpful to add some more context to the documentation of the methods (it is not obvious what saliency means and does).
There is quite some code duplication because of the distinction between ensemble and non-ensemble input arrays (but this already in the original version from @wdewettin I suppose). I don't have a solution yet, but I feel we should be able to improve this too.

Nice to have (and we can help with that): 
- [x] tests
- [x] example script

**Comment by RubenImhoff** [comment]
Great work, Nathalie! I'll have a look at it soon. :)

File: pysteps/blending/linear_blending.py (hunk 4 above)
Comment (dnerini) [reply pysteps/blending/linear_blending.py:212]: I think we should be able to use the solution from https://stackoverflow.com/a/24399139/4222370 , something like
```
# 
if n_ens_members_max == 1:
   ref_dim = 0
else:
   ref_dim = 1

# compute the slice indices
def _get_slice(n_dims, ref_dim, ref_id):
  slc = [slice(None)] * n_dims
  slc[ref_dim] = ref_id
  return tuple(slc)

# apply blending
slc_id= _get_slice(precip_blended.ndim, ref_dim, i)
precip_blended[slc_id] = ws * precip_nowcast[slc_id] + (1 - ws) * precip_nwp[slc_id]
```

File: pysteps/blending/interface.py [hunk 13]
```diff
@@ -32,6 +35,13 @@ def get_method(name):
     | linear_blending  | the linear blending of a nowcast method with other   |
     |                  | data (e.g. NWP data).                                |
     +------------------+------------------------------------------------------+
+    | salient_blending | the salient blending of a nowcast method with other  |
+    |                  | data (e.g. NWP data) described in :cite:`Hwang2015`. |
+    |                  | The blending is based on intensities and forecast    |
+    |                  | times. The blended product preserve pixel            |
```
Comment (RubenImhoff) [inline pysteps/blending/interface.py]:
```suggestion
    |                  | times. The blended product preserves pixel            |
```

File: pysteps/blending/interface.py [hunk 14]
```diff
@@ -32,6 +35,13 @@ def get_method(name):
...
+    |                  | intensities with time if they are strong enough based|
+    |                  | on there ranked salience.                            |
```
Comment (RubenImhoff) [inline pysteps/blending/interface.py]:
```suggestion
    |                  | on their ranked salience.                            |
```

File: pysteps/blending/linear_blending.py [hunk 15]
```diff
@@ -22,6 +22,7 @@
 import numpy as np
```
Comment (RubenImhoff) [inline pysteps/blending/linear_blending.py:23]: Perhaps we should add some explanation about the saliency-based blending method in the docstring above (so the first lines of the code), too. This focusses only on the linear blending method, now, but you have of course nicely extended the method with the saliency-based blending approach.

File: pysteps/blending/linear_blending.py [hunk 16]
```diff
@@ -83,6 +85,11 @@ def forecast(
       no data mask of the nowcast.
     nowcast_kwargs: dict, optional
       Dictionary containing keyword arguments for the nowcast method.
+    saliency: bool, optional
+      Standard value is False. If True, saliency will be used for blending. The blending
+      is based on intensities and forecast times as described in :cite:`Hwang2015`. The blended
+      product preserve pixel intensities with time if they are strong enoug based on there ranked
```
Comment (RubenImhoff) [inline pysteps/blending/linear_blending.py]:
```suggestion
      product preserve pixel intensities with time if they are strong enoug based on their ranked
```

File: pysteps/blending/linear_blending.py [hunk 17]
```diff
@@ -37,6 +38,7 @@ def forecast(
     end_blending=240,
     fill_nwp=True,
     nowcast_kwargs=None,
+    saliency=False,
```
Comment (RubenImhoff) [inline pysteps/blending/linear_blending.py:41]: Perhaps we should put this line above the 'nowcast_kwargs=None' line - but that is more from a 'cleaner' code perspective.

File: pysteps/blending/linear_blending.py [hunk 18]
```diff
@@ -37,6 +38,7 @@ def forecast(
...
 ):
 
     """Generate a forecast by linearly blending nowcasts with NWP data
```
Comment (RubenImhoff) [inline pysteps/blending/linear_blending.py]:
```suggestion
    """Generate a forecast by linearly or saliency-based blending of nowcasts and NWP data
```

File: pysteps/blending/linear_blending.py [hunk 19]
```diff
@@ -166,45 +173,137 @@ def forecast(
...
+                ref_dim = 0
+            else:
+                ref_dim = 1
+
+            # compute the slice indices
+            def _get_slice(n_dims, ref_dim, ref_id):
+                """source: https://stackoverflow.com/a/24399139/4222370"""
+                slc = [slice(None)] * n_dims
+                slc[ref_dim] = ref_id
+                return tuple(slc)
```
Comment (RubenImhoff) [inline pysteps/blending/linear_blending.py]: Maybe nicer to put this function at the end of the code (similar to the steps.py code).

File: pysteps/blending/interface.py [hunk 20]
```diff
@@ -11,11 +11,14 @@
...
 """
 
+from functools import partial
+
 from pysteps.blending import linear_blending
 from pysteps.blending import steps
 
 _blending_methods = dict()
 _blending_methods["linear_blending"] = linear_blending.forecast
+_blending_methods["salient_blending"] = partial(linear_blending.forecast, saliency=True)
```
Comment (RubenImhoff) [inline pysteps/blending/interface.py:21]: Cool, was not aware of this tool, useful!

**Review by RubenImhoff — CHANGES_REQUESTED** [review]
Hi Nathalie,

Really nice work! I like the saliency-based blending method and it seems like an excellent addition to the existing blending methods. 

As this method is added to the linear_blending method (makes sense), I think we can 'simply' extend the tests. Daniele and I surely can help with that, indeed. 

Perhaps we should change the name linear_blending.py to something else (no suggestion ready yet, haha, I'm thinking about it)? In that case, we would have to rename the tests and parts of the interface as well. What do you think?

File: pysteps/blending/linear_blending.py (hunk 4 above)
Comment (RubenImhoff) [reply pysteps/blending/linear_blending.py:212]: This is indeed an elegant solution!

File: pysteps/blending/linear_blending.py (hunk 3 above)
Comment (dnerini) [reply pysteps/blending/linear_blending.py]: thanks! it remains only to change it too in the `Returns` block in the docstring

File: pysteps/blending/linear_blending.py (hunk 5 above)
Comment (NathalieRombeek) [reply pysteps/blending/linear_blending.py]: In the paper they use different weights to calculate ws. However, following the approach of the paper, the values of the salient weight won't add up to 1. It is now changed by only calculating ws once.

File: examples/plot_linear_blending.py [hunk 21]
```diff
@@ -166,17 +191,27 @@
 for n, leadtime in enumerate(leadtimes_min):
 
     # Nowcast with blending into NWP
-    plt.subplot(n_leadtimes, 2, n * 2 + 1)
+    plt.subplot(n_leadtimes, 3, n * 2 + 1)
```
Comment (dnerini) [inline examples/plot_linear_blending.py]:
```suggestion
    plt.subplot(n_leadtimes, 3, n * 3 + 1)
```

File: examples/plot_linear_blending.py [hunk 22]
```diff
@@ -166,17 +191,27 @@
...
         precip_blended[int(leadtime / timestep) - 1, :, :],
         geodata=radar_metadata,
-        title=f"Nowcast +{leadtime} min",
+        title=f"Linear +{leadtime} min",
+        axis="off",
+        colorbar=False,
+    )
+
+    # Nowcast with salient blending into NWP
+    plt.subplot(n_leadtimes, 3, n * 2 + 2)
```
Comment (dnerini) [inline examples/plot_linear_blending.py]:
```suggestion
    plt.subplot(n_leadtimes, 3, n * 3 + 2)
```

File: examples/plot_linear_blending.py [hunk 23]
```diff
@@ -166,17 +191,27 @@
...
+        precip_salient_blended[int(leadtime / timestep) - 1, :, :],
+        geodata=radar_metadata,
+        title=f"Salient +{leadtime} min",
         axis="off",
         colorbar=False,
     )
 
     # Raw NWP forecast
-    plt.subplot(n_leadtimes, 2, n * 2 + 2)
+    plt.subplot(n_leadtimes, 3, n * 2 + 3)
```
Comment (dnerini) [inline examples/plot_linear_blending.py]:
```suggestion
    plt.subplot(n_leadtimes, 3, n * 3 + 3)
```

File: examples/plot_linear_blending.py [hunk 24]
```diff
@@ -148,16 +148,41 @@
...
 # extrapolation nowcast). Between 60 and 120 min, the NWP forecast gradually gets more
-# weight, whereas the extrapolation forecasts gradually gets less weight.
-# After 120 min, the blended forecast entirely consists of the NWP rainfall
-# forecast.
+# weight, whereas the extrapolation forecasts gradually gets less weight. In addition,
+# the saliency based blending takes also pixel intensities into account, which are
+# preserved over time. After 120 min, the blended forecast entirely consists of the
+# NWP rainfall forecast.
 
 fig = plt.figure(figsize=(4, 8))
```
Comment (dnerini) [inline examples/plot_linear_blending.py]: did you try to run the example? is the original figsize sufficient even with an additional column?

File: pysteps/blending/interface.py [hunk 25]
```diff
@@ -32,6 +35,13 @@ def get_method(name):
     | linear_blending  | the linear blending of a nowcast method with other   |
     |                  | data (e.g. NWP data).                                |
     +------------------+------------------------------------------------------+
+    | salient_blending | the salient blending of a nowcast method with other  |
+    |                  | data (e.g. NWP data) described in :cite:`Hwang2015`. |
+    |                  | The blending is based on intensities and forecast    |
+    |                  | times. The blended product preserves pixel            |
```
Comment (dnerini) [inline pysteps/blending/interface.py]:
```suggestion
    |                  | times. The blended product preserves pixel           |
```

File: pysteps/blending/linear_blending.py [hunk 26]
```diff
@@ -81,12 +84,18 @@ def forecast(
     fill_nwp: bool, optional
       Standard value is True. If True, the NWP data will be used to fill in the
       no data mask of the nowcast.
+    saliency: bool, optional
+      Standard value is False. If True, saliency will be used for blending. The blending
```
Comment (dnerini) [inline pysteps/blending/linear_blending.py]:
```suggestion
      Default value is False. If True, saliency will be used for blending. The blending
```

File: examples/plot_linear_blending.py [hunk 27]
```diff
@@ -148,16 +148,41 @@
 )
 
 
+################################################################################
+# The salient blending of nowcast and NWP rainfall forecast
+# ~~~~~~~~~~~~~~~~~~~~
```
Comment (dnerini) [inline examples/plot_linear_blending.py]:
```suggestion
# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

File: pysteps/blending/linear_blending.py [hunk 28]
```diff
@@ -9,9 +9,10 @@
 consists of the nowcast(s). In between the start and end time, the nowcast(s)
 weight decreases and NWP forecasts weight increases linearly from 1(0) to
 0(1). After the end time, the blended forecast entirely consists of the NWP
-forecasts.
+forecasts. The saliency-based blending method also takes into account the pixel
```
Comment (dnerini) [inline pysteps/blending/linear_blending.py]:
```suggestion
forecasts. The saliency-based blending method also takes into account the pixel
```

**Review by dnerini — CHANGES_REQUESTED** [review]
Thanks @NathalieRombeek for this new version, I feel we're getting very close to merging ;-)

I only left few super minor suggestions and comments, after this I'd be happy to approve this PR!

File: examples/plot_linear_blending.py [hunk 29]
```diff
@@ -148,16 +148,41 @@
...
+#
+# This method follows the procedure described in :cite:`Hwang2015`. The blending is
```
Comment (RubenImhoff) [inline examples/plot_linear_blending.py]:
```suggestion
# This method follows the saliency-based blending procedure described in :cite:`Hwang2015`. The blending is
```

File: examples/plot_linear_blending.py [hunk 30]
```diff
@@ -148,16 +148,41 @@
...
 # ~~~~~~~~~~~~~~~~~~~~
 #
 # The linear blending starts at 60 min, so during the first 60 minutes the
 # blended forecast only consists of the extrapolation forecast (consisting of an
 # extrapolation nowcast). Between 60 and 120 min, the NWP forecast gradually gets more
-# weight, whereas the extrapolation forecasts gradually gets less weight.
-# After 120 min, the blended forecast entirely consists of the NWP rainfall
-# forecast.
+# weight, whereas the extrapolation forecasts gradually gets less weight. In addition,
+# the saliency based blending takes also pixel intensities into account, which are
```
Comment (RubenImhoff) [inline examples/plot_linear_blending.py]:
```suggestion
# the saliency-based blending takes also pixel intensities into account, which are
```

File: examples/plot_linear_blending.py [hunk 31]
```diff
@@ -166,17 +191,27 @@
 for n, leadtime in enumerate(leadtimes_min):
 
     # Nowcast with blending into NWP
-    plt.subplot(n_leadtimes, 2, n * 2 + 1)
+    plt.subplot(n_leadtimes, 3, n * 2 + 1)
     plot_precip_field(
         precip_blended[int(leadtime / timestep) - 1, :, :],
         geodata=radar_metadata,
-        title=f"Nowcast +{leadtime} min",
+        title=f"Linear +{leadtime} min",
```
Comment (RubenImhoff) [inline examples/plot_linear_blending.py]:
```suggestion
        title=f"Linear + {leadtime} min",
```

File: examples/plot_linear_blending.py [hunk 32]
```diff
@@ -166,17 +191,27 @@
...
+        axis="off",
+        colorbar=False,
+    )
+
+    # Nowcast with salient blending into NWP
+    plt.subplot(n_leadtimes, 3, n * 2 + 2)
+    plot_precip_field(
+        precip_salient_blended[int(leadtime / timestep) - 1, :, :],
+        geodata=radar_metadata,
+        title=f"Salient +{leadtime} min",
```
Comment (RubenImhoff) [inline examples/plot_linear_blending.py]:
```suggestion
        title=f"Salient + {leadtime} min",
```

File: examples/plot_linear_blending.py [hunk 33]
```diff
@@ -148,16 +148,41 @@
...
 # The linear blending starts at 60 min, so during the first 60 minutes the
 # blended forecast only consists of the extrapolation forecast (consisting of an
 # extrapolation nowcast). Between 60 and 120 min, the NWP forecast gradually gets more
-# weight, whereas the extrapolation forecasts gradually gets less weight.
-# After 120 min, the blended forecast entirely consists of the NWP rainfall
-# forecast.
+# weight, whereas the extrapolation forecasts gradually gets less weight. In addition,
+# the saliency based blending takes also pixel intensities into account, which are
+# preserved over time. After 120 min, the blended forecast entirely consists of the
+# NWP rainfall forecast.
```
Comment (RubenImhoff) [inline examples/plot_linear_blending.py]: Perhaps also briefly write something about the difference we see between the two methods.

**Review by RubenImhoff — APPROVED** [review]
Great work, Nathalie! 
I have mentioned just a few minor suggestions. Other than that, it is good to go!

File: examples/plot_linear_blending.py (hunk 24 above)
Comment (NathalieRombeek) [reply examples/plot_linear_blending.py]: Yes, I also checked the example now and changed the figsize!

File: examples/plot_linear_blending.py (hunk 33 above)
Comment (NathalieRombeek) [reply examples/plot_linear_blending.py]: Thank you for your suggestions! I added a short explanation about the salient blending and added the figures of the extrapolation forecasts to make it more clear.

**Review by dnerini — APPROVED** [review]

**Comment by dnerini** [comment]
Congratulations @NathalieRombeek on your first and excellent contribution to pysteps!

**Pull request merged**